Bottom: 6ea4bc3f83688bf5e95d1de3dc08f1a96c146eeb
Top:    00f466efa1a8832981dfcebee031e824a6c103ed
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2008-10-11 22:44:47 +0100

Use the r_trees to avoid needing to check so many polygons all the time


---

diff --git a/src/find.c b/src/find.c
index 76cdf21..7b6507f 100644
--- a/src/find.c
+++ b/src/find.c
@@ -33,8 +33,6 @@
  * - lists for pins and vias, lines, arcs, pads and for polygons are created.
  *   Every object that has to be checked is added to its list.
  *   Coarse searching is accomplished with the data rtrees.
- * - there's no 'speed-up' mechanism for polygons because they are not used
- *   as often as other objects 
  * - the maximum distance between line and pin ... would depend on the angle
  *   between them. To speed up computation the limit is set to one half
  *   of the thickness of the objects (cause of square pins).
@@ -2427,6 +2425,29 @@ LOCtoPolyRat_callback (const BoxType * b, void *cl)
   return 0;
 }
 
+static int
+LOCtoPolyPolygon_callback (const BoxType * b, void *cl)
+{
+  PolygonTypePtr polygon = (PolygonTypePtr) b;
+  struct lo_info *i = (struct lo_info *) cl;
+
+  if (!TEST_FLAG (TheFlag, polygon) &&
+      IsPolygonInPolygon (polygon, &i->polygon) &&
+      ADD_POLYGON_TO_LIST (i->layer, polygon))
+    longjmp (i->env, 1);
+
+  return 0;
+}
+
+static int
+LOCtoPolyPourPolygon_callback (const BoxType * b, void *cl)
+{
+  PourTypePtr pour = (PourTypePtr) b;
+  struct pv_info *i = (struct pv_info *) cl;
+
+  return r_search (pour->polygon_tree, (BoxType *) &i->pv,
+                   NULL, LOCtoPolyPolygon_callback, i);
+}
 
 /* ---------------------------------------------------------------------------
  * looks up LOs that are connected to the given polygon
@@ -2460,19 +2481,12 @@ LookupLOConnectionsToPolygon (PolygonTypePtr Polygon, Cardinal LayerGroup)
       if (layer < max_layer)
         {
           /* check all polygons */
-
-          POUR_LOOP (LAYER_PTR (layer));
-          {
-            POURPOLYGON_LOOP (pour);
-            {
-              if (!TEST_FLAG (TheFlag, polygon) &&
-                  IsPolygonInPolygon (polygon, Polygon) &&
-                  ADD_POLYGON_TO_LIST (layer, polygon))
-                return True;
-            }
-            END_LOOP;
-          }
-          END_LOOP;
+          if (setjmp (info.env) == 0)
+            r_search (LAYER_PTR (layer)->pour_tree,
+                      (BoxType *) & info.polygon, NULL,
+                      LOCtoPolyPourPolygon_callback, &info);
+          else
+            return True;
 
           info.layer = layer;
           /* check all lines */
